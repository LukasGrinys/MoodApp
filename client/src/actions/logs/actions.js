export const actions = {
    canLog : {
        request : 'CAN_LOG_REQUEST',
        receive : 'CAN_LOG_RECEIVE'
    },
    getLastLogs : {
        request : 'GET_LAST_LOGS_REQUEST',
        receive : 'GET_LAST_LOGS_RECEIVE',
        reset : 'GET_LAST_LOGS_RESET'
    },
    getLog : {
        request : 'GET_LOG_REQUEST',
        receive : 'GET_LOG_RECEIVE',
        reset : 'GET_LOG_RESET'
    },
    getLogs : {
        request : 'GET_LOGS_REQUEST',
        receive : 'GET_LOGS_RECEIVE',
        reset : 'GET_LOGS_RESET'
    },
    postLog : {
        request : 'POST_LOG_REQUEST',
        receive : 'POST_LOG_RECEIVE',
        reset : 'POST_LOG_RESET'
    }
}