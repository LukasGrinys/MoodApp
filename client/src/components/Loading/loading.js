import React from 'react';
import styles from './loading.module.css';
import { useTheme } from '../../contexts/ThemeContext';
import classNames from 'classnames'

const Loading = () => {
    const darkTheme = useTheme();

    return (
        <div className={styles.loadingScreen}>
            <div>
                Loading    
            </div>
            <br/>
            <div className={classNames(
                styles.barContainer,
                darkTheme && styles.dark
            )}>
                <div className={styles.bar}></div>
                <div className={[styles.bar, styles.delay1].join(' ')}></div>
                <div className={[styles.bar, styles.delay2].join(' ')}></div>
            </div>
        </div>
    );
};

export default Loading;