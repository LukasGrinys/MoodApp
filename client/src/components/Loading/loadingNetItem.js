import React from 'react';
import styles from './loadingNetItem.module.css';

const LoadingNetItem = () => {
    return (
        <div className={styles.loadingScreen}>
            <div className={styles.barContainer}>
                <div className={styles.bar}></div>
                <div className={[styles.bar, styles.delay1].join(' ')}></div>
                <div className={[styles.bar, styles.delay2].join(' ')}></div>
            </div>
        </div>
    );
};

export default LoadingNetItem;