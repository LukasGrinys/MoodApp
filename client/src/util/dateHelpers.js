import { daytimes } from '../constants/daytimes';

export const getDate = () => {
    const date = new Date();
    
    const month = date.getMonth() < 10 ? `0${date.getMonth() + 1}` : `${date.getMonth() + 1}`;  
    const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate();
    const fullDate = [date.getFullYear(), month, day].join('-');

    return fullDate;
}

export const getDaytime = () => {
    const date = new Date();

    const hour = date.getHours();

    if (hour >= 7 && hour <= 12) {
        return daytimes.morning
    }

    if (hour > 12 && hour <= 19) {
        return daytimes.afternoon
    }

    if (hour > 19 && hour <= 23) {
        return daytimes.evening
    }

    if (hour < 7) {
        return daytimes.night
    }

    return null
}