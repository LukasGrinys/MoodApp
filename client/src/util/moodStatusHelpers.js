import moodTypes from '../constants/moodTypes';

export const returnMoodText = unparsedRating => {
    const rating = parseInt(unparsedRating);

    const moodType = moodTypes.find( type => type.above < rating && type.below >= rating);

    if (!moodType) {
        return '';
    }

    return moodType.text
}

export const returnAverageRating = logs => {
    let total = 0;
    let count = 0;

    logs.forEach( log => {
        total += parseInt(log.rating);
        count++;
    });

    const average = Math.round(((total/count)*10)/10);

    if (Number.isInteger(average)) {
        return `${average}.0`
    }

    return String(average);
}