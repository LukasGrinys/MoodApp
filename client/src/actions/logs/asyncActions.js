import axios from 'axios';
import { actions } from './actions';

export const canUserLog = async (currentDate, currentTiming, user) => {
    const body = {
        userId: user,
        date: currentDate,
        timing: currentTiming
    };

    return async dispatch => {
        const { data } = await axios.post('/api/canLog', body);

        dispatch({type: actions.canLog.request})
        
        if (data && data.data) {
            dispatch({
                type: actions.canLog.receive,
                payload: data.data.canLog
            })

            return;
        }

        dispatch({
            type: actions.canLog.receive,
            payload: false
        })
    }
}

export const getLastLogs = async ({
    userId,
    limit = 9
}) => {
    const url = `/api/getLogs?id=${userId}&limit=${limit}`;

    return async dispatch => {
        dispatch({type: actions.getLastLogs.request});

        try {
            const { data } = await axios.get(url);
            const { success, error } = data;
    
            if (error) {
                dispatch({
                    type: actions.getLastLogs.receive,
                    payload: {
                        logs: [],
                        error: error,
                        isFetching: false
                    }
                });
    
                return;
            }
    
            if (success && data.data) {
                dispatch({
                    type: actions.getLastLogs.receive,
                    payload: {
                        logs: data.data,
                        error: null,
                        isFetching: false
                    }
                });
    
                return;
            }
        } catch (error) {
            dispatch({
                type: actions.getLastLogs.receive,
                payload: {
                    logs: [],
                    error: error,
                    isFetching: false
                }
            });
        }
    }
}

export const clearLastLogs = () => dispatch => dispatch({type: actions.getLastLogs.reset});

export const getLog = async logId => {
    return async dispatch => {
        const endpoint = `/api/getLog?id=${logId}`;

        dispatch({type: actions.getLog.request});

        try {
            const response = await axios.get(endpoint);
            
            const { error, data } = response.data;

            if (error) {
                dispatch({
                    type: actions.getLog.receive,
                    payload: {
                        isFetching : false,
                        error,
                        logData : {}
                    }
                });

                return;
            }
            
            dispatch({
                type: actions.getLog.receive,
                payload: {
                    isFetching : false,
                    error : false,
                    logData : data
                }
            });
        } catch (error) {
            dispatch({
                type: actions.getLog.receive,
                payload: {
                    isFetching : false,
                    error,
                    logData : {}
                }
            });
        }
    }
}

export const clearLog = () => dispatch => { dispatch({type: actions.getLog.reset}) }

export const getLogs = async ({userId, start, limit, currentLogs}) => {
    return async dispatch => {
        const endpoint = `/api/getLogs?id=${userId}${start ? `&skip=${start}` : ''}&limit=${limit}`;

        dispatch({ type: actions.getLogs.request });

        try {
            const response = await axios.get(endpoint);
            const { data, error } = response.data;

            if (error) {
                dispatch({
                    type: actions.getLogs.receive,
                    payload: {
                        isFetching: false,
                        error
                    }
                });

                return;
            }
            
            dispatch({
                type: actions.getLogs.receive,
                payload: {
                    isFetching: false,
                    error: false,
                    logs: currentLogs ? [...currentLogs, ...data] : data,
                    noLogsLeft: !data.length ? true : false
                }
            });
        } catch (error) {
            dispatch({
                type: actions.getLogs.receive,
                payload: {
                    isFetching: false,
                    error
                }
            });
        }
    }
}

export const clearLogs = () => dispatch => { dispatch({ type: actions.getLogs.reset }) }

export const postLog = async ({
    userId,
    date,
    timing,
    rating,
    text
}) => {
    const body = {
        userId,
        date,
        timing,
        rating,
        text
    }

    return async dispatch => {
        try {
            dispatch( {type: actions.postLog.request});
    
            const { data } = await axios.post('/api/postLog', body);
            const { error , success } = data;

            if (error || !success) {
                dispatch({
                    type: actions.postLog.receive,
                    payload: {
                        isPosting : false,
                        success : false,
                        error : error || 'Unknown error occured'
                    }
                });

                return;
            }

            dispatch({
                type: actions.postLog.receive,
                payload: {
                    isPosting : false,
                    success : true,
                    error : null
                }
            })
        } catch (error) {
            dispatch({
                type: actions.postLog.receive,
                payload: {
                    isPosting : false,
                    success : false,
                    error
                }
            })
        }
    }  
}

export const clearLogPost = () => dispatch => { dispatch( {type: actions.postLog.reset})};