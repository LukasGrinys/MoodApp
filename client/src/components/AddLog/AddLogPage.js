import React, { Fragment } from 'react';
import { Wrapper, Heading, Label, Input, Footer } from '../Forms/Forms';
import Button from '../Button';
import BackButton from '../Back';
import Loading from '../Loading/loading';
import ErrorBox from '../ErrorBox';
import FontAwesome from 'react-fontawesome';
import styles from './AddLogPage.module.scss';
import { useAddLog } from '../../hooks/AddLog/useAddLog';

const AddLogPage = () => {
    const {
        returnMoodText,
        postLogError,
        canLog,
        message,
        moodRating,
        isPostingLog,
        handleRangeInput,
        handleMessageInput,
        handleAddLog
    } = useAddLog();

    if (canLog === undefined) {
        return <Loading/>
    }

    if (canLog) {
        return (
            <Fragment>
                <Wrapper>
                    <Heading>Add log</Heading>
                    <span>How do you feel at the moment? Rate your mood:</span>
                    <div className={styles.topSection}>
                        <div className={styles.moodBox}>
                            {moodRating}
                        </div>
                        <div className={styles.inputLine}>
                            <FontAwesome 
                                className={styles.thumb} 
                                name="thumbs-down" 
                                style={{color: "#800000"}}
                            />
                            <input 
                                onChange={handleRangeInput} 
                                type="range" 
                                min="1" 
                                max="10" 
                                value={moodRating}
                            />
                            <FontAwesome 
                                className={styles.thumb} 
                                name="thumbs-up" 
                                style={{color: "#16774f"}}
                            />
                        </div>
                        <span>{returnMoodText(moodRating)}</span>
                    </div>
                    <Label>Describe your thoughts: (optional)</Label>
                    <Input 
                        type={'textarea'}
                        name={'message'}
                        value={message}
                        controlEvents={{
                            handleChange : handleMessageInput
                        }}
                    />
                    <Footer>
                        <Button
                            handleClick={handleAddLog}
                            disabled={isPostingLog}
                        >
                            Submit
                        </Button>
                    </Footer>
                    {postLogError && (
                            <ErrorBox>
                                {postLogError}
                            </ErrorBox>
                        )}
                </Wrapper>
                <BackButton/>
            </Fragment>
        );
    }

    return null;
};

export default AddLogPage;