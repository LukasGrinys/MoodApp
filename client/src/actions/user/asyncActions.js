import { actions } from './actions';
import axios from 'axios';
import { clearLogs, clearLastLogs } from '../logs/asyncActions';
import Cookie from 'universal-cookie';
const cookie = new Cookie();

export const createUser = async user => {
    return async dispatch => {
        dispatch({type: actions.createUser.request});

        try {
            const { data } = await axios.post('/api/users', user)
            const { success, error } = data;

            if (error || !success) {
                dispatch({
                    type: actions.createUser.receive,
                    payload: {
                        success: false,
                        error: error || 'Could not create user'
                    }
                });

                return;
            }
            
            if (success) {
                dispatch({
                    type: actions.createUser.receive,
                    payload: {
                        success: true,
                        error: null
                    }
                });
            }
        } catch (error) {
            dispatch({
                type: actions.createUser.receive,
                payload: {
                    success: false,
                    error: error || 'Could not create user'
                }
            });
        }
    }
}

export const clearCreateUser = () => dispatch => dispatch({type: actions.createUser.reset});

export const loginUser = async payload => {
    return async dispatch => {
        dispatch({
            type: actions.loginUser.request
        });

        try {
            const { data } = await axios.post('/api/login', payload);
            const { isAuth, error } = data;

            if (error) {
                dispatch({
                    type: actions.loginUser.receive,
                    payload: {
                        isAuth: false,
                        authError : error
                    }
                });

                return;
            }

            if (isAuth) {
                dispatch({
                    type: actions.loginUser.receive,
                    payload: {
                        isAuth,
                        error: null
                    }
                });
            }
        } catch (error) {
            dispatch({
                type: actions.loginUser.receive,
                payload: {
                    isAuth: false,
                    error
                }
            })
        }
    }
}

export function logOut() {
    const request = axios.get('/api/logout');

    return dispatch => {
        request.then( (response) => {
            if (response.status === 200) {
                cookie.remove('auth');
                
                dispatch({
                    type: actions.clearUserData
                });
                
                clearLogs();
                clearLastLogs();
            }
        })
    }
}

export function auth() {
    const response = axios.get('/api/auth')
    .then( ({data}) => data);

    return {
        type: actions.userAuth,
        payload: response
    }
}

export const editUserDetails = async ({userId, details}) => {
    const endpoint = '/api/editAccount';
    const body = {
        id : userId,
        ...details
    }

    return async dispatch => {
        dispatch({type: actions.editUser.request});

        try {
            const { data } = await axios.post(endpoint, body);
            const { error } = data;

            if (error) {
                dispatch({
                    type: actions.editUser.receive,
                    payload: {
                        isEditing: false,
                        success: false,
                        error: error.toString()
                    }
                });

                return
            }

            dispatch({
                type: actions.editUser.receive,
                payload: {
                    isEditing: false,
                    success: true,
                    error: null
                }
            });
        } catch (error) {
            dispatch({
                type: actions.editUser.receive,
                payload: {
                    isEditing: false,
                    success: false,
                    error: error.toString()
                }
            });
        }
    }
}

export const clearEditUser = () => dispatch => {dispatch({type: actions.editUser.reset})}

export const changePassword = async ({
    userId,
    currentPassword,
    newPassword
}) => {
    return async dispatch => {
        if (!userId || !currentPassword || !newPassword) {
            return;
        }

        const endpoint = '/api/changePassword';
        const body = {
            id : userId,
            oldPassword : currentPassword,
            newPassword
        }

        await dispatch({type: actions.changePassword.request});

        try {
            const { error } = await axios.post(endpoint, body);

            if (error) {
                dispatch({
                    type: actions.changePassword.receive,
                    payload: {
                        isChanging : false,
                        error : error,
                        success : false
                    }
                });

                return;
            }

            dispatch({
                type: actions.changePassword.receive,
                payload: {
                    isChanging : false,
                    error : null,
                    success : true
                }
            });
        } catch (error) {
            dispatch({
                type: actions.changePassword.receive,
                payload: {
                    isChanging : false,
                    error : error,
                    success : false
                }
            });
        }
    }
}

export const clearChangePassword = async () => async dispatch => {dispatch({type: actions.changePassword.reset})}

export const clearUserData = async () => async dispatch => {dispatch({type: actions.clearUserData})}