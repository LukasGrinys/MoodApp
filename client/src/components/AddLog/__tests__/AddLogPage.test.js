import React from 'react';
import AddLogPage from '../AddLogPage';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import Loading from '../../Loading/loading';
import { testUser } from '../../../../__mocks__/userMock';
import { postLog } from '../../../actions/logs/asyncActions';
import { getDate, getDaytime } from '../../../util/dateHelpers';

const mockStore = configureStore();
const mockDispatch = jest.fn();

jest.mock('react-router-dom', () => {
    return {
        useHistory: jest.fn()
    }
});

jest.mock("react-redux", () => {
    const { Provider, useSelector } = jest.requireActual("react-redux");
  
    return {
        useDispatch: () => mockDispatch,
        useSelector,
        Provider
    };
});


describe('AddLogPage: loading', () => {
    const store = mockStore({
        user: { userData: testUser },
        logs: { canLog: undefined }
    })
    const wrapper = mount(
        <Provider store={store}>
            <AddLogPage/>
        </Provider>
    );

    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has loading widget', () => {
        expect(wrapper.find(Loading).length).toBe(1);
    });
});

const store = mockStore({
    user: { userData: testUser },
    logs: { canLog: true }
});

const wrapper = mount(
    <Provider store={store}>
        <AddLogPage/>
    </Provider>
);

describe('AddLogPage UI', () => {
    test('Renders properly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('Has input of type "range" for mood rating', () => {
        expect(wrapper.find('input[type="range"]').length).toBe(1);
    });

    test('Input range element has proper initial props', () => {
        expect(wrapper.find('input[type="range"]').props()).toEqual({
            max: "10",
            min: "1",
            onChange: expect.any(Function),
            type: "range",
            value: 5
        })
    });

    test('Has a block for displaying mood text', () => {
        expect(wrapper.find('.topSection').children('span').length).toBe(1);
    })

    test('Has textarea for log message', () => {
        expect(wrapper.find('textarea').length).toBe(1);
    });

    test('Textarea has proper initial props', () => {
        expect(wrapper.find('textarea').props()).toEqual({
            className: 'textArea',
            name: 'message',
            onChange: expect.any(Function),
            placeholder: '',
            value: ''
        });
    });

    test('Has button for submitting log', () => {
        expect(wrapper.find('button').length).toBe(1);
    });

    test('Button has proper initial props', () => {
        expect(wrapper.find('button').props()).toEqual({
            className: "button btnPrimary",
            disabled: false,
            onClick: expect.any(Function),
            type: 'button',
            children: 'Submit'
        })
    });
});

describe('AddLogPage logic', () => {
    const testRating = 10;
    const testMessage = 'This is a test message';

    test('Mood rating changes when input range is changed', () => {
        wrapper.find('input[type="range"]').simulate('change', {
            target : {
                type: 'range',
                value: testRating
            }
        });

        expect(wrapper.find('.moodBox').text()).toBe(testRating.toString());
    });

    test('Mood text changes according to rating', () => {
        expect(wrapper.find('.topSection').children('span').text()).toBe('Awesome!');
    });

    test('Log message changes when typed to textarea', () => {
        wrapper.find('textarea').simulate('change', {
            target : {
                name: 'message',
                value: testMessage
            }
        });

        expect(wrapper.find('textarea').prop('value')).toBe(testMessage);
    });

    test('Button click dispatches postLog action', () => {
        const button = wrapper.find('button');
        button.simulate('click');

        setTimeout( () => {
            expect(postLog).toHaveBeenCalledTimes(1);

            expect(postLog).toHaveBeenCalledWidth({
                userId: testUser.userData.id,
                date : getDate(),
                timing : getDaytime(),
                rating : testRating,
                text : testMessage
            });

            done();
        }, 0);
    });

    test('Button should be disabled after submitting log', () => {
        expect(wrapper.find('button').prop('disabled')).toBeTruthy();
    });

    test('Pressing enter key submits log', () => {
        wrapper.find('textarea').simulate('keypress', {key: 'Enter'});

        setTimeout( () => {
            expect(postLog).toHaveBeenCalledTimes(1);
            done();
        }, 0)
    })
});