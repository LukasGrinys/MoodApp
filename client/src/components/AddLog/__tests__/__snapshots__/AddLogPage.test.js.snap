// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLogPage UI Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AddLogPage>
    <Wrapper>
      <div
        className="formWrapper"
      >
        <Heading>
          <h1
            className="heading"
          >
            Add log
          </h1>
        </Heading>
        <span>
          How do you feel at the moment? Rate your mood:
        </span>
        <div
          className="topSection"
        >
          <div
            className="moodBox"
          >
            5
          </div>
          <div
            className="inputLine"
          >
            <FontAwesome
              className="thumb"
              name="thumbs-down"
              style={
                Object {
                  "color": "#800000",
                }
              }
            >
              <span
                aria-hidden={true}
                className="fa fa-thumbs-down thumb"
                style={
                  Object {
                    "color": "#800000",
                  }
                }
              />
            </FontAwesome>
            <input
              max="10"
              min="1"
              onChange={[Function]}
              type="range"
              value={5}
            />
            <FontAwesome
              className="thumb"
              name="thumbs-up"
              style={
                Object {
                  "color": "#16774f",
                }
              }
            >
              <span
                aria-hidden={true}
                className="fa fa-thumbs-up thumb"
                style={
                  Object {
                    "color": "#16774f",
                  }
                }
              />
            </FontAwesome>
          </div>
          <span>
            Average
          </span>
        </div>
        <Label>
          <span
            className="label"
          >
            Describe your thoughts: (optional)
          </span>
        </Label>
        <Input
          controlEvents={
            Object {
              "handleChange": [Function],
            }
          }
          name="message"
          placeholder=""
          type="textarea"
          value=""
        >
          <textarea
            className="textArea"
            name="message"
            onChange={[Function]}
            placeholder=""
            value=""
          />
        </Input>
        <Footer>
          <div
            className="formFooter"
          >
            <Button
              disabled={false}
              handleClick={[Function]}
            >
              <button
                className="button btnPrimary"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Submit
              </button>
            </Button>
          </div>
        </Footer>
      </div>
    </Wrapper>
    <BackButton>
      <div
        className="container"
      >
        <div
          className="button"
          onClick={[Function]}
        >
          <FontAwesome
            name="arrow-left"
          >
            <span
              aria-hidden={true}
              className="fa fa-arrow-left"
            />
          </FontAwesome>
          <div
            className="text"
          >
            Go back
          </div>
        </div>
      </div>
    </BackButton>
  </AddLogPage>
</Provider>
`;

exports[`AddLogPage: loading Renders properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AddLogPage>
    <Loading>
      <div
        className="loadingScreen"
      >
        <div>
          Loading
        </div>
        <br />
        <div
          className="barContainer"
        >
          <div
            className="bar"
          />
          <div
            className="bar delay1"
          />
          <div
            className="bar delay2"
          />
        </div>
      </div>
    </Loading>
  </AddLogPage>
</Provider>
`;
